binary_search_tree/tree.py

```diff
@@ -14,44 +14,157 @@ class Tree:
+            return None
+
+        current = self.root
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif current.key < key:
+                current = current.right
+            else:
+                current = current.left
+        return None
+
+# IN ORDER LIST*************************************************************
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
+    def inorder_helper(self, current_node, inorder_list):
+        if current_node == None:
+            return inorder_list
+
+        # ADD LEFT NODE FIRST
+        self.inorder_helper(current_node.left, inorder_list)
+
+        # ADD CURRENT NODE
+        inorder_list.append({ 
+            "key": current_node.key,
+            "value": current_node.value
+        })
+        # ADD RIGHT NODE LAST
+        self.inorder_helper(current_node.right, inorder_list)
+
+        return inorder_list
 
-    # Time Complexity: 
-    # Space Complexity: 
     def inorder(self):
-        pass
+        inorder_list = []
+        return self.inorder_helper(self.root, inorder_list)
+
+
+# PRE-ORDER LIST*************************************************************
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)    
+    def preorder_helper(self, current_node, preorder_list):
+        if current_node == None:
+            return preorder_list
+
+        # ADD CURRENT NODE FIRST 
+        preorder_list.append({ 
+            "key": current_node.key,
+            "value": current_node.value
+        })
+        # ADD LEFT NODE, THEN RIGHT NODE
+        self.preorder_helper(current_node.left, preorder_list)
+        self.preorder_helper(current_node.right, preorder_list)
+
+        return preorder_list
 
-    # Time Complexity: 
-    # Space Complexity:     
     def preorder(self):
-        pass
+        preorder_list = []
+        return self.preorder_helper(self.root, preorder_list)
+
+
+
+# POST-ORDER LIST*************************************************************
+
+    # Time Complexity: O(log n)
+    # Space Complexity:  O(1)   
+    def postorder_helper(self, current_node, postorder_list):
+        if current_node == None:
+            return postorder_list
+
+        # ADD LEFT NODE, THEN RIGHT NODE 
+        self.postorder_helper(current_node.left, postorder_list)
+        self.postorder_helper(current_node.right, postorder_list)
+
+        # ADD CURRENT NODE LAST 
+        postorder_list.append({ 
+            "key": current_node.key,
+            "value": current_node.value
+        })
+        return postorder_list
 
-    # Time Complexity: 
-    # Space Complexity:     
     def postorder(self):
-        pass
+        postorder_list = []
+        return self.postorder_helper(self.root, postorder_list)
+
 
-    # Time Complexity: 
-    # Space Complexity:     
+
+
+    # Time Complexity: O(log n)
+    # Space Complexity:  O(1)   
+    def height_helper(self, current_node):
```

Comment:
👍 Since you are building a list here, the space complexity is O(log n) which can be written as O(h) where h is the height of the tree.